File: discord-bot-js/src/audio.js
/**
 * Audio helpers: WAV encoding, STT (Whisper), TTS (ElevenLabs/OpenAI), wake word detection
 */

// ─── WAV encoding ─────────────────────────────────────────────────────────────

/**
 * Convert 48kHz stereo PCM (i16 interleaved) captured from Discord voice into
 * a 16kHz mono WAV buffer suitable for Whisper.
 */
export function downsampleAndEncodeWav(pcmBuffer) {
  // pcmBuffer is a Node.js Buffer of raw i16 LE stereo 48kHz samples
  const sampleCount = pcmBuffer.length / 2; // number of i16 samples
  const monoDownsampled = [];

  // stereo pairs are L, R, L, R...
  // Downsample 48kHz→16kHz: take every 3rd pair (ratio = 48000/16000 = 3)
  for (let i = 0; i < sampleCount; i += 6) {
    // i and i+1 are one stereo pair; step by 6 to get every 3rd pair
    const l = pcmBuffer.readInt16LE(i * 2);
    const r = pcmBuffer.readInt16LE((i + 1) * 2);
    monoDownsampled.push((l + r) >> 1);
  }

  const pcm16k = Buffer.allocUnsafe(monoDownsampled.length * 2);
  for (let i = 0; i < monoDownsampled.length; i++) {
    pcm16k.writeInt16LE(monoDownsampled[i], i * 2);
  }

  return buildWavHeader(pcm16k, 16000, 1);
}

function buildWavHeader(pcmData, sampleRate, channels) {
  const dataSize = pcmData.length;
  const header = Buffer.allocUnsafe(44);
  header.write('RIFF', 0);
  header.writeUInt32LE(36 + dataSize, 4);
  header.write('WAVE', 8);
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16);                               // fmt chunk size
  header.writeUInt16LE(1, 20);                                // PCM format
  header.writeUInt16LE(channels, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(sampleRate * channels * 2, 28);        // byte rate
  header.writeUInt16LE(channels * 2, 32);                     // block align
  header.writeUInt16LE(16, 34);                               // bits per sample
  header.write('data', 36);
  header.writeUInt32LE(dataSize, 40);
  return Buffer.concat([header, pcmData]);
}

// ─── STT ──────────────────────────────────────────────────────────────────────

/**
 * Transcribe WAV audio via local Whisper server or OpenAI Whisper API.
 * Returns null if audio is too short or no provider is configured.
 */
export async function transcribeWav(wavBuffer) {
  // Require at least 0.5 seconds: 16kHz mono i16 = 16000*2 bytes + 44 header
  const MIN_BYTES = 16_044;
  if (wavBuffer.length < MIN_BYTES) {
    console.debug(`[STT] Audio too short (${wavBuffer.length} bytes < ${MIN_BYTES}), skipping`);
    return null;
  }

  console.log(`[STT] Transcribing ${wavBuffer.length} byte WAV (~${((wavBuffer.length - 44) / 32000).toFixed(1)}s audio)`);

  const whisperUrl = process.env.WHISPER_URL;
  if (whisperUrl) {
    try {
      // Try /transcribe/file (whisper_server.py) then /inference (whispercpp/other servers)
      for (const path of ['/transcribe/file', '/inference']) {
        const form = new FormData();
        form.append('file', new Blob([wavBuffer], { type: 'audio/wav' }), 'audio.wav');
        form.append('language', 'en');
        try {
          const res = await fetch(`${whisperUrl.replace(/\/$/, '')}${path}`, {
            method: 'POST',
            body: form,
            signal: AbortSignal.timeout(30_000),
          });
          if (res.ok) {
            const json = await res.json();
            const text = (json.text ?? json.transcript)?.trim();
            if (text) {
              console.log(`[STT] Local Whisper result: "${text.slice(0, 80)}"`);
              return text;
            }
            // Empty transcription — nothing heard, stop here
            return null;
          }
        } catch (pathErr) {
          if (path === '/transcribe/file') continue; // try next path
          throw pathErr;
        }
      }
    } catch (e) {
      console.warn('[STT] Local Whisper failed:', e.message);
    }
  }

  // OpenAI Whisper fallback — only if no local provider configured
  if (!whisperUrl) {
    const apiKey = process.env.OPENAI_API_KEY;
    if (apiKey) {
      try {
        const form = new FormData();
        form.append('file', new Blob([wavBuffer], { type: 'audio/wav' }), 'audio.wav');
        form.append('model', 'whisper-1');
        form.append('language', 'en');
        const res = await fetch('https://api.openai.com/v1/audio/transcriptions', {
          method: 'POST',
          headers: { Authorization: `Bearer ${apiKey}` },
          body: form,
        });
        if (res.ok) {
          const json = await res.json();
          const text = json.text?.trim() || null;
          console.log(`[STT] OpenAI Whisper result: "${text?.slice(0, 80) ?? '(empty)'}"`);
          return text;
        } else {
          const body = await res.text();
          console.warn(`[STT] OpenAI Whisper returned ${res.status}: ${body.slice(0, 200)}`);
        }
      } catch (e) {
        console.warn('[STT] OpenAI Whisper failed:', e.message);
      }
    }
  }

  return null;
}

// ─── Wake word detection ──────────────────────────────────────────────────────

const WAKE_WORDS = {
  nora: ['hey nora', 'ok nora', 'nora'],
  topsi: ['hey topsi', 'ok topsi', 'topsi'],
};

// Regex for phonetic Topsi variants — Whisper consistently mishears it as
// tulsi, tofsi, topsy, tapsi, topci, topzy, topsei, topsie, etc.
// Pattern: starts with T, 4-7 chars, ends with s+vowel/y or si/sy
const TOPSI_REGEX = /\bt[aou][a-z]{0,2}s[aeiouy]\b/i;

/**
 * Checks whether the transcript addresses the given agent.
 * Returns the text after the wake word (the actual command), or null if not addressed.
 */
export function detectWakeWord(text, agentName) {
  const lower = text.toLowerCase();

  // Regex fallback for Topsi — catches all phonetic mishearings automatically
  if (agentName.toLowerCase() === 'topsi') {
    const m = lower.match(TOPSI_REGEX);
    if (m) {
      const idx = m.index + m[0].length;
      const after = text.slice(idx).replace(/^[^a-z0-9]+/i, '').trim();
      return after || text.trim();
    }
  }

  const words = WAKE_WORDS[agentName.toLowerCase()] ?? [];
  for (const w of words) {
    const idx = lower.indexOf(w);
    if (idx !== -1) {
      const after = text.slice(idx + w.length).replace(/^[^a-z0-9]+/i, '').trim();
      return after || text.trim();
    }
  }
  return null;
}

// ─── Agent call ───────────────────────────────────────────────────────────────

export async function callAgent(serverPort, agentName, message, projectId, sessionId, participantCtx = '') {
  const endpoint = agentName.toLowerCase() === 'topsi' ? 'topsi' : 'nora';
  const identities = {
    nora: `You are Nora, PowerClub Global's Executive AI Agent. You are speaking via Discord voice. British English.`,
    topsi: `You are Topsi, PowerClub Global's technical AI agent. You are speaking via Discord voice.`,
  };
  const identity = identities[agentName.toLowerCase()] ?? `You are ${agentName}, a PowerClub Global AI agent.`;
  const contextPart = participantCtx ? ` ${participantCtx}` : '';
  const lengthGuide = agentName.toLowerCase() === 'topsi'
    ? 'No markdown. Respond at whatever length the user requests — if asked to read aloud, read the full text without summarising.'
    : 'Keep your final spoken response to 1-3 sentences, no markdown.';
  const prefix = `[DISCORD VOICE CALL — ${identity}${contextPart} Speaking aloud — ${lengthGuide} You MAY and SHOULD still call tools (search_web, fetch_web_page, render_page, etc.) when the user asks you to look something up, fetch a URL, or search the web — tool use is encouraged.]`;

  const res = await fetch(`http://127.0.0.1:${serverPort}/api/internal/${endpoint}/chat`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-Admin-Key': process.env.ADMIN_API_KEY ?? '',
    },
    body: JSON.stringify({
      message: `${prefix}\n\n${message}`,
      sessionId,
    }),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Agent returned ${res.status}: ${body}`);
  }

  const json = await res.json();
  const response = json.response ?? json.message ?? json.content ?? 'Understood.';
  console.log(`[${agentName}] Response: ${response.slice(0, 80)}`);
  return response;
}

// ─── TTS ──────────────────────────────────────────────────────────────────────

/**
 * Synthesize text to an MP3 buffer.
 * Tries ElevenLabs first, falls back to OpenAI TTS.
 */
// Nora uses ElevenLabs (Mia Moore — British Studio Presenter)
// Topsi uses Chatterbox (local, matches dashboard config)
const ELEVENLABS_NORA_VOICE = process.env.ELEVENLABS_NORA_VOICE_ID ?? 'ZtcPZrt9K4w8e1OB9M6w';
const CHATTERBOX_URL = process.env.CHATTERBOX_URL ?? 'http://localhost:8102';

export async function synthesizeTts(text, agentName = 'nora') {
  const agent = agentName.toLowerCase();

  // ── Topsi: Piper TTS (en_GB-semaine-medium, prudence speaker) ────────────
  // Matches the dashboard voice: computer-sounding semaine female voice
  if (agent === 'topsi') {
    try {
      const res = await fetch(`${CHATTERBOX_URL.replace(/\/$/, '')}/tts`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          text,
          speaker_id: 'prudence',
          speed: 1.0,
        }),
        signal: AbortSignal.timeout(30_000),
      });
      if (res.ok) {
        console.log('[TTS] Piper semaine/prudence (Topsi)');
        return Buffer.from(await res.arrayBuffer());
      }
      console.warn(`[TTS] Piper returned ${res.status}`);
    } catch (e) {
      console.warn('[TTS] Piper failed:', e.message);
    }
    throw new Error('Piper TTS unavailable for Topsi — is it running on port 8102?');
  }

  // ── Nora: ElevenLabs (Mia Moore) ─────────────────────────────────────────
  const elKey = process.env.ELEVENLABS_API_KEY;
  if (elKey) {
    try {
      const res = await fetch(`https://api.elevenlabs.io/v1/text-to-speech/${ELEVENLABS_NORA_VOICE}`, {
        method: 'POST',
        headers: {
          'xi-api-key': elKey,
          'Content-Type': 'application/json',
          Accept: 'audio/mpeg',
        },
        body: JSON.stringify({
          text,
          model_id: 'eleven_monolingual_v1',
          voice_settings: { stability: 0.5, similarity_boost: 0.75 },
        }),
      });
      if (res.ok) {
        console.log('[TTS] ElevenLabs Mia Moore (Nora)');
        return Buffer.from(await res.arrayBuffer());
      }
      console.warn(`[TTS] ElevenLabs returned ${res.status}`);
    } catch (e) {
      console.warn('[TTS] ElevenLabs failed:', e.message);
    }
  }

  throw new Error('No TTS provider available for Nora — check ELEVENLABS_API_KEY');
}

File: crates/discord-bot/src/audio.rs
//! Audio processing utilities: PCM accumulation, WAV encoding, STT, wake word detection, TTS

use std::{io::Cursor, sync::Arc};

use anyhow::{Context, Result};
use tokio::sync::Mutex;
use tracing::{debug, info, warn};

use crate::session::ActiveAgent;

/// Encode 48kHz stereo i16 PCM as a WAV file in memory
pub fn encode_wav(stereo_samples: &[i16]) -> Result<Vec<u8>> {
    // Convert stereo to mono by averaging channels
    let mono: Vec<i16> = stereo_samples
        .chunks_exact(2)
        .map(|ch| {
            let l = ch[0] as i32;
            let r = ch[1] as i32;
            ((l + r) / 2) as i16
        })
        .collect();

    // Downsample from 48kHz to 16kHz (Whisper's preferred rate)
    // Simple decimation: take every 3rd sample (48000 / 16000 = 3)
    let downsampled: Vec<i16> = mono.iter().step_by(3).copied().collect();

    let mut wav_bytes = Vec::new();
    let cursor = Cursor::new(&mut wav_bytes);

    let spec = hound::WavSpec {
        channels: 1,
        sample_rate: 16000,
        bits_per_sample: 16,
        sample_format: hound::SampleFormat::Int,
    };

    let mut writer = hound::WavWriter::new(cursor, spec).context("Failed to create WAV writer")?;

    for sample in &downsampled {
        writer
            .write_sample(*sample)
            .context("Failed to write WAV sample")?;
    }
    writer.finalize().context("Failed to finalize WAV")?;

    Ok(wav_bytes)
}

/// Encode TTS audio (16kHz mono WAV) back to 48kHz stereo for Discord playback
pub fn upsample_to_discord(wav_bytes: &[u8]) -> Result<Vec<i16>> {
    let cursor = Cursor::new(wav_bytes);
    let mut reader = hound::WavReader::new(cursor).context("Failed to open TTS WAV")?;

    let spec = reader.spec();
    let samples: Vec<i16> = reader
        .samples::<i16>()
        .collect::<std::result::Result<Vec<_>, _>>()
        .context("Failed to read TTS samples")?;

    // Upsample from source rate to 48kHz
    let src_rate = spec.sample_rate as f64;
    let dst_rate = 48000.0_f64;
    let ratio = dst_rate / src_rate;
    let dst_len = (samples.len() as f64 * ratio) as usize;

    let mut upsampled = Vec::with_capacity(dst_len * 2); // stereo

    for i in 0..dst_len {
        let src_idx = i as f64 / ratio;
        let lo = src_idx.floor() as usize;
        let hi = (lo + 1).min(samples.len() - 1);
        let frac = src_idx - lo as f64;

        let lo_val = samples[lo] as f64;
        let hi_val = samples[hi] as f64;
        let sample = (lo_val + (hi_val - lo_val) * frac) as i16;

        // Duplicate to stereo L + R
        upsampled.push(sample);
        upsampled.push(sample);
    }

    Ok(upsampled)
}

/// Transcribe WAV audio via OpenAI Whisper API or local Whisper server
pub async fn transcribe_wav(wav_bytes: Vec<u8>) -> Result<Option<String>> {
    // Require at least 0.5 seconds of audio (16kHz mono = 8000 samples = 16000 bytes + 44 header)
    if wav_bytes.len() < 16_044 {
        debug!("Audio too short to transcribe ({} bytes)", wav_bytes.len());
        return Ok(None);
    }

    // Try local Whisper server first
    if let Ok(whisper_url) = std::env::var("WHISPER_URL") {
        match transcribe_local(&whisper_url, wav_bytes.clone()).await {
            Ok(text) if !text.trim().is_empty() => return Ok(Some(text)),
            Ok(_) => {}
            Err(e) => warn!("Local Whisper failed: {}", e),
        }
    }

    // Fall back to OpenAI Whisper API
    if let Ok(api_key) = std::env::var("OPENAI_API_KEY") {
        match transcribe_openai(&api_key, wav_bytes).await {
            Ok(text) if !text.trim().is_empty() => return Ok(Some(text)),
            Ok(_) => {}
            Err(e) => warn!("OpenAI Whisper failed: {}", e),
        }
    }

    warn!("No STT provider available — set WHISPER_URL or OPENAI_API_KEY");
    Ok(None)
}

async fn transcribe_local(whisper_url: &str, wav_bytes: Vec<u8>) -> Result<String> {
    let client = reqwest::Client::new();
    let url = format!("{}/inference", whisper_url.trim_end_matches('/'));

    let part = reqwest::multipart::Part::bytes(wav_bytes)
        .file_name("audio.wav")
        .mime_str("audio/wav")?;
    let form = reqwest::multipart::Form::new().part("file", part);

    let resp = client
        .post(&url)
        .multipart(form)
        .send()
        .await
        .context("Local Whisper request failed")?;

    let json: serde_json::Value = resp.json().await?;
    Ok(json["text"].as_str().unwrap_or("").to_string())
}

async fn transcribe_openai(api_key: &str, wav_bytes: Vec<u8>) -> Result<String> {
    let client = reqwest::Client::new();

    let part = reqwest::multipart::Part::bytes(wav_bytes)
        .file_name("audio.wav")
        .mime_str("audio/wav")?;
    let form = reqwest::multipart::Form::new()
        .part("file", part)
        .text("model", "whisper-1")
        .text("language", "en");

    let resp = client
        .post("https://api.openai.com/v1/audio/transcriptions")
        .bearer_auth(api_key)
        .multipart(form)
        .send()
        .await
        .context("OpenAI Whisper request failed")?;

    let json: serde_json::Value = resp.json().await?;
    Ok(json["text"].as_str().unwrap_or("").to_string())
}

/// Check if the transcript text addresses the given agent
/// Returns the addressed text (after the wake word) if detected
pub fn detect_wake_word(text: &str, agent: ActiveAgent) -> Option<String> {
    let lower = text.to_lowercase();

    for pattern in agent.wake_words() {
        if let Some(pos) = lower.find(pattern) {
            let after = pos + pattern.len();
            let remainder = if after < text.len() {
                text[after..]
                    .trim_start_matches(|c: char| !c.is_alphabetic() && !c.is_numeric())
                    .trim()
                    .to_string()
            } else {
                String::new()
            };

            return Some(remainder);
        }
    }
    None
}

/// Call the PCG server agent chat endpoint and return the response text
pub async fn call_agent(
    server_port: u16,
    agent: ActiveAgent,
    user_message: &str,
    project_id: &str,
    session_id: &str,
) -> Result<String> {
    let client = reqwest::Client::new();
    let base_url = format!("http://127.0.0.1:{}/api", server_port);

    let voice_prefix = format!(
        "[DISCORD VOICE CALL — Speaking to {}. Voice only. 1-3 sentences max. No markdown. British English.]",
        agent.name()
    );
    let full_message = format!("{}\n\n{}", voice_prefix, user_message);

    let body = serde_json::json!({
        "message": full_message,
        "projectId": project_id,
        "sessionId": session_id,
    });

    let endpoint = match agent {
        ActiveAgent::Nora => format!("{}/nora/chat", base_url),
        ActiveAgent::Topsi => format!("{}/topsi/chat", base_url),
    };

    // Use the internal API key for service-to-service calls
    let api_key = std::env::var("ADMIN_API_KEY").unwrap_or_default();

    let resp = client
        .post(&endpoint)
        .header("X-Internal-Service", "discord-bot")
        .header("X-Admin-Key", &api_key)
        .json(&body)
        .send()
        .await
        .context("Agent call failed")?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        anyhow::bail!("Agent returned {}: {}", status, text);
    }

    let json: serde_json::Value = resp.json().await?;
    let response_text = json["response"]
        .as_str()
        .or_else(|| json["message"].as_str())
        .or_else(|| json["content"].as_str())
        .unwrap_or("Understood.")
        .to_string();

    info!(
        "Agent {} responded: {}",
        agent.name(),
        &response_text[..response_text.len().min(80)]
    );
    Ok(response_text)
}

/// Synthesize text to WAV bytes via configured TTS provider
pub async fn synthesize_tts(text: &str) -> Result<Vec<u8>> {
    // Try ElevenLabs first
    if let Ok(api_key) = std::env::var("ELEVENLABS_API_KEY") {
        let voice_id = std::env::var("ELEVENLABS_VOICE_ID")
            .unwrap_or_else(|_| "21m00Tcm4TlvDq8ikWAM".to_string()); // Rachel voice

        match synthesize_elevenlabs(&api_key, &voice_id, text).await {
            Ok(wav) => return Ok(wav),
            Err(e) => warn!("ElevenLabs TTS failed: {}", e),
        }
    }

    // Fall back to OpenAI TTS
    if let Ok(api_key) = std::env::var("OPENAI_API_KEY") {
        match synthesize_openai(&api_key, text).await {
            Ok(wav) => return Ok(wav),
            Err(e) => warn!("OpenAI TTS failed: {}", e),
        }
    }

    anyhow::bail!("No TTS provider available — set ELEVENLABS_API_KEY or OPENAI_API_KEY")
}

async fn synthesize_elevenlabs(api_key: &str, voice_id: &str, text: &str) -> Result<Vec<u8>> {
    let client = reqwest::Client::new();
    let url = format!(
        "https://api.elevenlabs.io/v1/text-to-speech/{}/stream",
        voice_id
    );

    let body = serde_json::json!({
        "text": text,
        "model_id": "eleven_monolingual_v1",
        "voice_settings": {
            "stability": 0.5,
            "similarity_boost": 0.75
        },
        "output_format": "pcm_16000"
    });

    let resp = client
        .post(&url)
        .header("xi-api-key", api_key)
        .header("Accept", "audio/mpeg")
        .json(&body)
        .send()
        .await?;

    if !resp.status().is_success() {
        anyhow::bail!("ElevenLabs returned {}", resp.status());
    }

    let audio_bytes = resp.bytes().await?.to_vec();

    // ElevenLabs pcm_16000 returns raw 16-bit LE PCM at 16kHz mono
    // Wrap in a WAV header so our upsample function can read it
    let mut wav_bytes = Vec::new();
    let cursor = Cursor::new(&mut wav_bytes);
    let spec = hound::WavSpec {
        channels: 1,
        sample_rate: 16000,
        bits_per_sample: 16,
        sample_format: hound::SampleFormat::Int,
    };
    let mut writer = hound::WavWriter::new(cursor, spec)?;
    for chunk in audio_bytes.chunks_exact(2) {
        let sample = i16::from_le_bytes([chunk[0], chunk[1]]);
        writer.write_sample(sample)?;
    }
    writer.finalize()?;

    Ok(wav_bytes)
}

async fn synthesize_openai(api_key: &str, text: &str) -> Result<Vec<u8>> {
    let client = reqwest::Client::new();

    let body = serde_json::json!({
        "model": "tts-1",
        "input": text,
        "voice": "nova",
        "response_format": "wav",
    });

    let resp = client
        .post("https://api.openai.com/v1/audio/speech")
        .bearer_auth(api_key)
        .json(&body)
        .send()
        .await?;

    if !resp.status().is_success() {
        anyhow::bail!("OpenAI TTS returned {}", resp.status());
    }

    Ok(resp.bytes().await?.to_vec())
}

/// Synthesize TTS and play it back into the Discord voice channel via songbird.
///
/// Writes audio to a temp WAV file, loads it as a songbird Input, and plays it.
/// The temp file is cleaned up after a short delay.
pub async fn synthesize_and_play(call_lock: Arc<Mutex<songbird::Call>>, text: &str) -> Result<()> {
    let wav_bytes = synthesize_tts(text).await?;

    // Write to a temp file so songbird's symphonia decoder can read it
    let tmp_path = format!("/tmp/pcg_tts_{}.wav", uuid::Uuid::new_v4());
    tokio::fs::write(&tmp_path, &wav_bytes)
        .await
        .context("Failed to write TTS temp file")?;

    let input = songbird::input::File::new(tmp_path.clone());

    {
        let mut handler = call_lock.lock().await;
        handler.play_input(input.into());
    }

    // Clean up temp file after the audio has had time to play
    let path = tmp_path.clone();
    tokio::spawn(async move {
        tokio::time::sleep(std::time::Duration::from_secs(30)).await;
        let _ = tokio::fs::remove_file(&path).await;
    });

    Ok(())
}
